### PR TITLE
Viewability

### PR DESCRIPTION
**Pull request by terryc33x**
add the viewaibility feature and updated the docs

### CODE REVIEW

File: thirty-three-across/DOCUMENTATION.md [hunk 1]
```diff
@@ -1,75 +1,179 @@
...
 | Key | Required | Type | Description |
 |---|---|---|---|
-| | | | |
- 
+|id|Yes|String| Bid request identifier |
+|imp|Yes|Object| ORTB imp object describes the information of an ad placement/slot |
+|site|Yes|Object| ORTB site object describes the information of publisher id, page, and ref |
+|user|No|Object| ORTB user object describes the information of user consent |
+|regs|No|Object| ORTB regs object describes the information about GDPR eligibility |
+|ext|No|Object| Prebid wrapper name and version info |
```
Comment (curlyblueeagle) [inline thirty-three-across/DOCUMENTATION.md]: Should the doc say `HB wrapper name and version info`?

File: thirty-three-across/DOCUMENTATION.md [hunk 2]
```diff
@@ -1,75 +1,179 @@
...
+
 ## Configuration Information
 ### Configuration Keys
 | Key | Required | Type | Description |
 |---|---|---|---|
-| | | | |
+|siteId|Yes|string| Publisher/site identifier |
+|productId|Yes|string| 33Across Product type, 'siab' or 'inview'  |
+|sizes|Yes| Array[] | An array of array representation of the size of the ad unit, like [[width1, height1],[width2, height2]] |
+|bidfloor|No|number| Minimum bid for this impression expressed in CPM |
```
Comment (curlyblueeagle) [inline thirty-three-across/DOCUMENTATION.md]: CPM in dollars or cents?

File: thirty-three-across/thirty-three-across-htb.js [hunk 3]
```diff
@@ -56,40 +57,176 @@ function ThirtyThreeAcrossHtb(configs) {
...
+            }
+        }
+
+        bbox.width = bbox.right - bbox.left;
+        bbox.height = bbox.bottom - bbox.top;
+
+        return bbox;
+    }
+
+    function _getPercentInView(element, topWin, minSize) {
```
Comment (curlyblueeagle) [inline thirty-three-across/thirty-three-across-htb.js:137]: Can we explain in function header comments how view-ability is measured?

File: thirty-three-across/thirty-three-across-htb-system-tests.js [hunk 4]
```diff
@@ -56,6 +56,10 @@ function validateBidRequest(request) {
         w: sizes[1][0],
         h: sizes[1][1]
     });
+    expect(queryObj.imp[0].banner.ext).toBeDefined();
+    expect(queryObj.imp[0].banner.ext.ttx).toBeDefined();
+    expect(queryObj.imp[0].banner.ext.ttx.viewability).toBeDefined();
+    expect(queryObj.imp[0].banner.ext.ttx.viewability.amount).toBeDefined();
```
Comment (curlyblueeagle) [inline thirty-three-across/thirty-three-across-htb-system-tests.js:62]: Can we not test this more by mocking element dimensions etc anc checking for actual values in different scenarios?

File: thirty-three-across/thirty-three-across-htb.js [hunk 5]
```diff
@@ -197,7 +334,7 @@ function ThirtyThreeAcrossHtb(configs) {
         if (ComplianceService.isPrivacyEnabled()) {
             var gdprStatus = ComplianceService.gdpr.getConsent();
 
-            gdprConsent = Object.assign(gdprConsent, {
+            gdprConsent = Utilities.mergeObjects(gdprConsent, {
```
Comment (curlyblueeagle) [inline thirty-three-across/thirty-three-across-htb.js:341]: Interesting, why the switch to `Utilities.mergeObjects`?

File: thirty-three-across/thirty-three-across-htb.js [hunk 6]
```diff
@@ -56,40 +57,176 @@ function ThirtyThreeAcrossHtb(configs) {
...
+         *      return true;
+         *  }
+         */
+        return !Browser.isTopFrame();
+    }
+
+    function _getAdSlotHTMLElement(parcel) {
+        var slotElementId = null;
+
+        // NOTE: viewability feature supports the DFP tag only for now
```
Comment (curlyblueeagle) [inline thirty-three-across/thirty-three-across-htb.js:181]: If it's not GPT then do we return `nm`?

**Review by curlyblueeagle — COMMENTED** [review]
I have left some comments/questions for more clarification. Thank you.

File: thirty-three-across/thirty-three-across-htb-system-tests.js (hunk 4 above)
Comment (terryc33x) [reply thirty-three-across/thirty-three-across-htb-system-tests.js:62]: we can not really do this, one is these viewability measurement are heavily depends on the DOM, and mocking the DOM is not easy. second is this test case is not control by us, we can not run this test case for different scenarios

File: thirty-three-across/thirty-three-across-htb.js (hunk 5 above)
Comment (terryc33x) [reply thirty-three-across/thirty-three-across-htb.js:341]: for IE9 compatibility. `Utilities.mergeObjects` is their util function, and its supporting IE9.

File: thirty-three-across/thirty-three-across-htb.js (hunk 6 above)
Comment (terryc33x) [reply thirty-three-across/thirty-three-across-htb.js:181]: yes. currently thats what prebid does, so i guess IX adapter should follow it too

File: thirty-three-across/thirty-three-across-htb-system-tests.js (hunk 4 above)
Comment (curlyblueeagle) [reply thirty-three-across/thirty-three-across-htb-system-tests.js:62]: Hmm in Prebid we tested like so https://github.com/33Across/Prebid.js/blob/34cd13384033c5ee8d54d552c621f463325e6696/test/spec/modules/33acrossBidAdapter_spec.js#L146

Not asking you to mock the DOM. Asking you to mock the dimensions returned by your helper functions.

**Pull request merged**